Web Inspector: the frontend should be able to prevent the inspected page from clearing the console

dcrousso opened the pull request:
#### 80d190d733833b4730f5acdf7ccd8eb6ef98227d
<pre>
Web Inspector: the frontend should be able to prevent the inspected page from clearing the console
<a href="https://bugs.webkit.org/show_bug.cgi?id=270753">https://bugs.webkit.org/show_bug.cgi?id=270753</a>

Reviewed by Patrick Angle.

&quot;malicious&quot; pages can sometimes repeatedly call `console.clear()` to prevent logging from being readable by the developer.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createConsoleSettingsView):
Add a checkbox in the Console pane of the Settings Tab to control this.

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.initializeTarget): Added.
(WI.ConsoleManager.prototype.messagesCleared):
(WI.ConsoleManager.prototype._setConsoleClearAPIEnabled): Added.
(WI.ConsoleManager.prototype._handleConsoleClearAPIEnabledSettingChanged): Added.
* Source/WebInspectorUI/UserInterface/Protocol/Target.js:
(WI.Target.prototype.initialize):
Notify all `WI.Target` whenever the above checkbox changes (and when Web Inspector is first opened).

* Source/JavaScriptCore/inspector/protocol/Console.json:
* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.cpp:
(Inspector::InspectorConsoleAgent::clearMessages):
(Inspector::InspectorConsoleAgent::setConsoleClearAPIEnabled): Added.
(Inspector::InspectorConsoleAgent::addMessageToConsole):
(Inspector::InspectorConsoleAgent::addConsoleMessage):
(Inspector::isGroupMessage): Deleted.
Listen for the above checkbox changes and prevent clearing existing `ConsoleMessages` if desired (which also means `Console.messagesCleared` is not sent).

* Source/JavaScriptCore/inspector/ConsoleMessage.cpp:
(Inspector::isGroupMessage): Added.
(Inspector::ConsoleMessage::isEqual const):
Previously, any `console.clear()` would remove all `ConsoleMessage`, meaning that there&apos;d never be a way for it to have a repeat count other than 1.  Now, `console.clear()` might not do that (depending on the above checkbox in the frontend), so make sure the backend treats all `console.clear()` as unique.
Drive-by: do the same for `console.group()`, `console.groupCollapsed()`, and `console.groupEnd()`.

* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.messageAdded):
Drive-by: make sure to ignore all `console.clear()` regardless of reason.

* Source/WebCore/inspector/CommandLineAPIHost.idl:
* Source/WebCore/inspector/CommandLineAPIHost.h:
* Source/WebCore/inspector/CommandLineAPIHost.cpp:
(WebCore::CommandLineAPIHost::clearConsoleMessages): Deleted.
Drive-by: remove unused code.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/console/setConsoleClearAPIEnabled.html: Added.
* LayoutTests/inspector/console/setConsoleClearAPIEnabled-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/276362@main">https://commits.webkit.org/276362@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6497dccf5013d0207e905f75aed554cf989a0f4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43888 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46528 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39965 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46192 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26820 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20334 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36210 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44463 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19958 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37792 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17212 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17480 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38873 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1940 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/37405 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40049 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39172 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48109 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/43484 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18892 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15855 "Found 1 new test failure: http/tests/navigation/parsed-iframe-dynamic-form-back-entry.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43042 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20285 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38063 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41747 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9887 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20489 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50516 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19909 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10194 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on LayoutTests/inspector/console/setConsoleClearAPIEnabled.html:
```diff
@@ -0,0 +1,111 @@
+            await InspectorTest.evaluateInPage(`console.clear()`);
+
+            InspectorTest.expectFalse(cleared, "Should not clear console.");
+            InspectorTest.expectFalse(messageAdded, "Should not send a message.");
```
What guarantees that the event handlers on the frontend _did not run_ and that's why the flags both `false`?

dcrousso replied:
oh hmm good point.  ill add a `console.log(42);` or something after the `console.clear()` to make sure

rcaliman-apple commented on LayoutTests/inspector/console/setConsoleClearAPIEnabled.html:
```diff
@@ -0,0 +1,111 @@
+            await InspectorTest.evaluateInPage(`console.clear()`);
+
+            InspectorTest.expectTrue(cleared, "Should clear console.");
+            InspectorTest.expectFalse(messageAdded, "Should not send a message.");
```
Same question here: what ensures the event listeners did run by the time these assertions are made?

dcrousso replied:
FWIW we should never dispatch `WI.ConsoleManager.Event.MessageAdded` when receiving a `Console.messageAdded` as we have a special `Console.messagesCleared` for that reason

rcaliman-apple replied:
My question is more about what prevents a race condition in this test now or in the future: the assertions running before the event handlers change the boolean flags?

dcrousso replied:
FWIW we have assumptions like this _all_ over Web Inspector code since our protocol (which is backed by XPC) is guaranteed to deliver messages in the order that they're sent

in this case ive added a `console.log()` after the `console.clear()` to ensure that a `Console.messageAdded` comes after a `Console.messageCleared`

rcaliman-apple replied:
Fair enough. At first glance this looks vulnerable to race conditions, but I'm satisfied with the explanation about the guaranteed order of messages over the protocol.

rcaliman-apple commented on Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.h:
```diff
@@ -99,6 +100,7 @@ class JS_EXPORT_PRIVATE InspectorConsoleAgent : public InspectorAgentBase, publi
     HashMap<String, MonotonicTime> m_times;
     bool m_enabled { false };
     bool m_isAddingMessageToFrontend { false };
+    bool m_consoleClearAPIEnabled { false };
```
Is there a reason for this flag to be default `false` while the frontend setting is default `true`?

dcrousso replied:
eh i generally err on the side of the default value being `false` since it keeps the memory as `0`.  i could rework this to be `m_consoleClearAPIDisabled` but generally across Web Inspector we prefer `*Enabled`

rcaliman-apple replied:
The naming is good 👍, no need to change that. 

It's just a bit odd that the backend starts in a default `false` state and expects the frontend to flip it to `true` which is the _real_ default state. Someone reading this later might be confused and question whether there's some subtle behavior being accounted for.

I'd suggest keeping the two in sync for clarity.

rcaliman-apple reviewed (commented):
Looking really nice and clean! I have just two questions. Everything else LGTM.

patrickangle reviewed (approved):
lgtm as well

dcrousso committed:
Web Inspector: the frontend should be able to prevent the inspected page from clearing the console
https://bugs.webkit.org/show_bug.cgi?id=270753

Reviewed by Patrick Angle.

"malicious" pages can sometimes repeatedly call `console.clear()` to prevent logging from being readable by the developer.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createConsoleSettingsView):
Add a checkbox in the Console pane of the Settings Tab to control this.

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.initializeTarget): Added.
(WI.ConsoleManager.prototype.messagesCleared):
(WI.ConsoleManager.prototype._setConsoleClearAPIEnabled): Added.
(WI.ConsoleManager.prototype._handleConsoleClearAPIEnabledSettingChanged): Added.
* Source/WebInspectorUI/UserInterface/Protocol/Target.js:
(WI.Target.prototype.initialize):
Notify all `WI.Target` whenever the above checkbox changes (and when Web Inspector is first opened).

* Source/JavaScriptCore/inspector/protocol/Console.json:
* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.cpp:
(Inspector::InspectorConsoleAgent::clearMessages):
(Inspector::InspectorConsoleAgent::setConsoleClearAPIEnabled): Added.
(Inspector::InspectorConsoleAgent::addMessageToConsole):
(Inspector::InspectorConsoleAgent::addConsoleMessage):
(Inspector::isGroupMessage): Deleted.
Listen for the above checkbox changes and prevent clearing existing `ConsoleMessages` if desired (which also means `Console.messagesCleared` is not sent).

* Source/JavaScriptCore/inspector/ConsoleMessage.cpp:
(Inspector::isGroupMessage): Added.
(Inspector::ConsoleMessage::isEqual const):
Previously, any `console.clear()` would remove all `ConsoleMessage`, meaning that there'd never be a way for it to have a repeat count other than 1.  Now, `console.clear()` might not do that (depending on the above checkbox in the frontend), so make sure the backend treats all `console.clear()` as unique.
Drive-by: do the same for `console.group()`, `console.groupCollapsed()`, and `console.groupEnd()`.

* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.messageAdded):
Drive-by: make sure to ignore all `console.clear()` regardless of reason.

* Source/WebCore/inspector/CommandLineAPIHost.idl:
* Source/WebCore/inspector/CommandLineAPIHost.h:
* Source/WebCore/inspector/CommandLineAPIHost.cpp:
(WebCore::CommandLineAPIHost::clearConsoleMessages): Deleted.
Drive-by: remove unused code.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/console/setConsoleClearAPIEnabled.html: Added.
* LayoutTests/inspector/console/setConsoleClearAPIEnabled-expected.txt: Added.

Canonical link: https://commits.webkit.org/276362@main
